Data/IntSet/Internal.hs

```diff
@@ -793,18 +796,30 @@ minView t =
 -- | /O(min(n,W))/. Delete and find the minimal element.
 --
 -- > deleteFindMin set = (findMin set, deleteMin set)
+#if __GLASGOW_HASKELL__ >= 800
+deleteFindMin :: HasCallStack => IntSet -> (Key, IntSet)
+#else
 deleteFindMin :: IntSet -> (Key, IntSet)
+#endif
 deleteFindMin = fromMaybe (error "deleteFindMin: empty set has no minimal element") . minView
 
 -- | /O(min(n,W))/. Delete and find the maximal element.
 --
 -- > deleteFindMax set = (findMax set, deleteMax set)
+#if __GLASGOW_HASKELL__ >= 800
+deleteFindMax :: HasCallStack => IntSet -> (Key, IntSet)
+#else
 deleteFindMax :: IntSet -> (Key, IntSet)
+#endif
 deleteFindMax = fromMaybe (error "deleteFindMax: empty set has no maximal element") . maxView
 
 
 -- | /O(min(n,W))/. The minimal element of the set.
+#if __GLASGOW_HASKELL__ >= 800
+findMin :: HasCallStack => IntSet -> Key
+#else
 findMin :: IntSet -> Key
+#endif
 findMin Nil = error "findMin: empty set has no minimal element"
```

Comment:
Does the call stack grow through the recursion here or elsewhere? If so, that's a big problem. You can check the Core to be sure. We know there are no `Nil`s except at the root, but GHC does not! If the stacks build, you'll need to restructure the functions to fix that. Watch out for performance. If the times for the current benchmarks exercising these functions are too short to trust, consider adding more benchmarks.

Reply:
Wow, yes it does.  I didn't realize it would do that.  The fix is to use an internal `go` function.  I'll update the PR with that in a moment.

On a related note, I realized that my `HasCallStack` for `Map.!` was wrong because `Map.!` calls `find` which actually throws the error.  Is there a reason it uses `find` rather than `lookup`?  My inclination is to remove `find` altogether (it's only used by `!` and looks identical other than the `Maybe` wrapper) and replace with `lookup`---then `!` can call `error` itself, which should then make `HasCallStack` work the way we want.

Reply:
Oh, I just noticed the `[Note: Local 'go' functions and capturing]` -- I'll read that and try to make sure I'm not doing anything stupid.